### DISCUSSION TITLE
Workaround for Linux screen reader

### DISCUSSION THREAD

**Discussion by PermanentNovice** [AvaloniaUI/Avalonia#14478] [General]
I am blind and have started coding a simple way to make Avalonia accessible with spd-say, a text-to-speech utility included with Gnome. In my app, I use Process.Start to launch an instance of spd-say and write text to the process’ input stream. My workaround is inefficient and ugly, but it is better than nothing and acts as a passable screen reader for my personal use.

In the simplest case, I created a label and set the IsTabStop and Focusable properties to true. I subscribe to the GotFocus event and in the handler method I pass the label’s text property to the spd-say process. Now, when I navigate to that label with the tab key, I hear the contents of the label using Linux, just like I do in Windows where Avalonia already has some built-in accessibility support.

In a more complex scenario involving a TextBox, I wrote logic in the backspace key up and down events and now the app reads characters deleted when I press backspace. This actually makes my Linux app more accessible than when run in Windows, because the TextBox control is not yet accessible, even for Window’s screen readers.

I wondered if there is a better way than subscribing to keypress and focus events and duplicating work that you may have already done. For Windows, I assume you collect the text that should be spoken before passing it to the accessibility interface. If I write code that detects the app is running in Linux, can I somehow get that text and pass it to my spd-say process?

**Comment by timunie** [comment]
@PermanentNovice on Linux it is hard for us to add support for screen readers, probably because of lack of knowledge. So if you can help us here to improve it, that would be very much appreciated. Please see: #14275

P.S.: Sorry, I do not have a better word for "Please see" to point to another issue. If anyone knows of a better wording here, please let me know.

**Reply by thevortexcloud** [reply]
How about "peruse", or simply "Check out"?

**Reply by PermanentNovice** [reply]
I followed that link and found that discussion is well beyond my skill level. I don’t know how a screen reader works, I just know how to use it to navigate through controls in an accessible application.

My workaround somewhat mimicked the simplest functionality of a screen reader. I know enough about c# to subscribe to an event and I passed some text to an external speech program. Since some work has been done for Windows to gather the text to be spoken, I had hoped there was an event raised before the text is passed to the native accessibility layer. If I can get that text, I can pass it to my ridiculous workaround.

P.S. I would have said "Please see" as well.

**Reply by khaledmahmoud78** [reply]
> How about "peruse", or simply "Check out"?

What is "peruse","?
Please provide us a link for it!

**Reply by thevortexcloud** [reply]
It's a slightly obscure English word.

>**peruse something** to read something, especially in a careful way
>* A copy of the report is available for you to peruse at your leisure.

https://www.oxfordlearnersdictionaries.com/definition/english/peruse

**Reply by khaledmahmoud78** [reply]
Thanks for your explanation!
